### ISSUE TITLE
Pitest: examine and try to remove infinite loop mutations

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#5673]
https://github.com/checkstyle/checkstyle/blob/master/pom.xml#L215-L216
Whenever pitest creates an infinite loop mutation, it takes a whole 50 seconds before it recognizes that the code will never finish. This slows down the pitest run considerably. Pitest can't recognize an infinite loop, and if the mutated code isn't an infinite loop, but takes a long time to finish, it can cause issues with mutation scoring.
I ran into this problem within my own project. An `uncertain` mutation is flagged and counted as a killed. However, if it is the edge of finishing before the timeout and does so in some circumstances, it could be flagged as `survived` and lower the score creating different results in different runs.

https://app.shippable.com/github/rnveach/checkstyle/runs/3/summary/console
Lists all the jobs who have uncertainties as `failed`.

It is possible to remove the uncertain instances by rewriting the code some.
Here is 1 example where I changed it from uncertain to survived: TBD

**Comment by romani** [comment]
It might be better to reduce 50 sec to 5 sec. Checkstyle tests are very quick, even 5 sec is a lot of time.
We can take a look at history of this parameter, it might be in such value from the beginning, then I copied draft config from my non open source project

**Comment by rnveach** [comment]
> We can take a look at history of this parameter, it might be in such value from the beginning

https://github.com/checkstyle/checkstyle/commit/ba178a2bb1aff1799d81742842a9b48ba93c8556#diff-600376dffeb79835ede4a0b285078036
It looks like it has been since the beginning.

**Comment by romani** [comment]
Let's try update to 5 sec first

**Comment by romani** [comment]
Execution in CI is critical for us, I would keep it in 2-3 time bigger than it takes on CI now.
We do not trust execution on local. If one test execution is above 5 sec and system is over used, there might be OOM or, .... .

**Comment by vivek-0509** [comment]
Hi, @romani  I am looking into this issue.

My understanding from the discussion is that simply reducing the timeout to 5 seconds is not the desired fix because it causes flakiness on slower machines. Instead, we should aim to refactor the code to eliminate the infinite loop mutations entirely, converting them into cleaner KILLED  outcomes.

I have started analyzing the pitest-coding-1 profile and found several TIMED_OUT mutations (e.g., in DeclarationOrderCheck). I was able to fix them by refactoring the manual Deque/while loop traversal into a safe hybrid iterative/recursive approach. This method iterates over siblings (preventing StackOverflowError on wide structures) and recurses on children, which effectively removes the fragile while loops that were causing infinite hangs during mutation.

My proposed plan is:

- Go through each Pitest profile one by one.
- Generate the report and identify all TIMED_OUT mutations.
- Refactor the underlying code to be robust against infinite loops 

Please let me know if this aligns with the expectations, or if there is a preferred alternative direction.

**Comment by romani** [comment]
please keep send ing PRs , evil is in details, please do single mutation in separate PR.

**Comment by MiniPiku** [comment]
Hi @romani  ,
I understand the preference for addressing mutations individually. In most cases that makes sense, but in few areas the affected mutants share the same structural root cause.

In `DeclarationOrderCheck.getAllTokensOfType()`, the previous implementation used a nested `while` loop with intertwined termination conditions:

```
while (vertex != null || !stack.isEmpty()) {  // outer loop
    if (!stack.isEmpty()) {
        vertex = stack.pop();
    }
    while (vertex != null) {  // inner loop
        ...
        vertex = vertex.getFirstChild();
    }
}
```
The unstable mutants (removed `isEmpty`, removed `pop`, negated conditionals, etc.) all exploit this same control-flow structure. Because the inner loop mutates vertex and the outer loop termination depends on both vertex and stack, any mutation to these conditions or stack operations leads to infinite traversal.

In other words, the 7 mutations are not independent defects , they are all symptoms of the same nested-loop structure. Fixing them individually would require modifying the same code block repeatedly, effectively reintroducing or partially reverting the structural issue.

For that reason, I addressed the shared root cause in a single localized change to the traversal logic. I will continue submitting separate PRs for any remaining independent mutations.

**Comment by romani** [comment]
All details on fix is at https://github.com/checkstyle/checkstyle/pull/19024
Done by @MiniPiku.
